### PR TITLE
API: Dynamically create clusterproviders (with caching)

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

Re-Enables #3904. The reason the initial version resulted in huge delays in the API is that during the ctrlruntimeclient creation, it does api discovery, which basically means it asks the target cluster what apis it has and traverses them down. There are quite a lot of apicalls done for that, for cloud asia this took 10-20S on my laptop.

This PR changes the whole thing to cache the discovery information and also warm up that cache during startup. If a seed gets added later on, the first request onto that seed will again take long.

I've used a `sync.Map` as this basically introduces a lock over all access to any seed and the info in the cache is something we very rarely have to update.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

/assign @maciaszczykm @floreks

### CODE REVIEW

**Comment by maciaszczykm** [comment]
Looks good, I will verify locally.

**Comment by maciaszczykm** [comment]
The problem seems to be fixed, UI works.

/approve

**Comment by maciaszczykm** [comment]
@floreks leaving to you for LGTM.

File: api/cmd/kubermatic-api/main.go [hunk 1]
```diff
@@ -357,3 +338,61 @@ func createAPIHandler(options serverRunOptions, prov providers, oidcIssuerVerifi
...
 	return instrumentHandler(mainRouter, lookupRoute), nil
 }
+
+func clusterProviderFactory(seedKubeconfigGetter provider.SeedKubeconfigGetter, workerName string, oidcKubeCfgEndpointEnabled bool) provider.ClusterProviderGetter {
+	// Discovery can take very long when the cluster is not physically close, so cache the discovery info
+	// use a sync.Map as we read a lot from this map but almost never write to it
+	restMapperCache := &sync.Map{}
+
+	return func(seedName string) (provider.ClusterProvider, error) {
+		cfg, err := seedKubeconfigGetter(seedName)
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-api/main.go:348]: `seedKubeconfigGetter` -> `getSeedKubeconfig`

File: api/cmd/kubermatic-api/main.go [hunk 2]
```diff
@@ -176,6 +137,26 @@ func createInitProviders(options serverRunOptions) (providers, error) {
...
 	}
+	seedKubeconfigGetter, err := provider.SeedKubeconfigGetterFactory(context.Background(), mgr.GetClient(), options.kubeconfig, options.dynamicDatacenters)
+	if err != nil {
+		return providers{}, err
+	}
+	clusterProviderGetter := clusterProviderFactory(seedKubeconfigGetter, options.workerName, options.featureGates.Enabled(OIDCKubeCfgEndpoint))
+
+	// Warm up the restMapper cache. Log but ignore errors encountered here, maybe there are stale seeds
+	go func() {
+		seeds, err := seedsGetter()
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-api/main.go:148]: `seedsGetter` -> `getSeeds`

**Comment by floreks** [comment]
Also tested locally. Seems to work fine.

/lgtm

File: api/cmd/kubermatic-api/main.go [hunk 3]
```diff
@@ -357,3 +338,61 @@ func createAPIHandler(options serverRunOptions, prov providers, oidcIssuerVerifi
...
+		}
+		kubeClient, err := kubernetes.NewForConfig(cfg)
+		if err != nil {
+			return nil, fmt.Errorf("faild to create kubeClient: %v", err)
+		}
+		defaultImpersonationClientForSeed := kubernetesprovider.NewKubermaticImpersonationClient(cfg)
+
+		var mapper meta.RESTMapper
+		mapperKey := fmt.Sprintf("%s/%s/%s/%s/%s/%s/%s/%s/%s/%s",
+			seedName, cfg.Host, cfg.APIPath, cfg.Username, cfg.Password, cfg.BearerToken, cfg.BearerTokenFile,
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-api/main.go:361]: Why is `seedName` not enough?

**Review by mrIncompetent — CHANGES_REQUESTED** [review]

File: api/cmd/kubermatic-api/main.go (hunk 3 above)
Comment (alvaroaleman) [reply api/cmd/kubermatic-api/main.go:361]: Because someone could create a new seed with the same name

File: api/cmd/kubermatic-api/main.go (hunk 1 above)
Comment (alvaroaleman) [reply api/cmd/kubermatic-api/main.go:348]: I don't really see how that makes anything more readable and its consistent with how its done everywhere else

File: api/cmd/kubermatic-api/main.go (hunk 2 above)
Comment (alvaroaleman) [reply api/cmd/kubermatic-api/main.go:148]: I don't really see how that makes anything more readable and its consistent with how its done everywhere else

File: api/cmd/kubermatic-api/main.go (hunk 3 above)
Comment (mrIncompetent) [reply api/cmd/kubermatic-api/main.go:361]: So add a comment explaining that

Comment (alvaroaleman) [reply api/cmd/kubermatic-api/main.go:361]: Done

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**